core-bundle/src/Image/Studio/FigureBuilder.php

```diff
@@ -0,0 +1,630 @@
+use Contao\FilesModel;
+use Contao\Image\ImageInterface;
+use Contao\Image\PictureConfiguration;
+use Contao\PageModel;
+use Contao\Validator;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\PathUtil\Path;
+
+/**
+ * Use the FigureBuilder class to create Figure result objects. The class
+ * has a fluent interface to configure the desired output. When you are ready,
+ * call build() to get a Figure. If you need another instance with similar
+ * settings, you can alter values and call build() again - it will not affect
+ * your first instance.
+ */
+class FigureBuilder
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $locator;
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * The resource's absolute file path.
+     *
+     * @var string|null
+     */
+    private $filePath;
+
+    /**
+     * The resource's file model if applicable.
+     *
+     * @var FilesModel|null
+     */
+    private $filesModel;
+
+    /**
+     * User defined size configuration.
+     *
+     * @phpcsSuppress SlevomatCodingStandard.Classes.UnusedPrivateElements
+     *
+     * @var mixed|null
+     */
+    private $sizeConfiguration;
+
+    /**
+     * User defined custom locale. This will overwrite the default if set.
+     *
+     * @var string|null
+     */
+    private $locale;
+
+    /**
+     * User defined meta data. This will overwrite the default if set.
+     *
+     * @var MetaData|null
+     */
+    private $metaData;
+
+    /**
+     * Determines if a meta data should never be present in the output.
+     *
+     * @var bool
+     */
+    private $disableMetaData;
+
+    /**
+     * User defined link attributes. These will add to or overwrite the default values.
+     *
+     * @var array<string, string|null>
+     */
+    private $additionalLinkAttributes = [];
+
+    /**
+     * User defined light box resource or url. This will overwrite the default if set.
+     *
+     * @var string|ImageInterface|null
+     */
+    private $lightBoxResourceOrUrl;
+
+    /**
+     * User defined light box size configuration. This will overwrite the default if set.
+     *
+     * @var mixed|null
+     */
+    private $lightBoxSizeConfiguration;
+
+    /**
+     * User defined light box group identifier. This will overwrite the default if set.
+     *
+     * @var string|null
+     */
+    private $lightBoxGroupIdentifier;
+
+    /**
+     * Determines if a light box (or "fullsize") image should be created.
+     *
+     * @var bool
+     */
+    private $enableLightBox;
+
+    /**
+     * User defined template options.
+     *
+     * @phpcsSuppress SlevomatCodingStandard.Classes.UnusedPrivateElements
+     *
+     * @var array<string, mixed>
+     */
+    private $options = [];
+
+    /**
+     * @internal Use the Contao\Image\Studio\Studio factory to get an instance of this class
+     */
+    public function __construct(ContainerInterface $locator)
```

Comment:
This still seems weird to me. Why do we not inject the required services and parameters directly?

```php
public function __construct(Studio $studio, ContaoFramework $framework, array $validExtensions, string $projectDir, string $uploadPath)
```

Then we would not have to suppress PhpStan errors, either.

```neon
        # Ignore that services via a locator are falsely identified as private/not registered
        - message: '#Service ".*" is not registered in the container\.#'
          paths:
              - %currentWorkingDirectory%/core-bundle/src/Image/Studio/*.php

        - message: '#Service "parameter_bag" is private\.#'
          paths:
              - %currentWorkingDirectory%/core-bundle/src/Image/Studio/*.php
```

Reply:
The service subscriber pattern allows lazy instantiation of services. Imo that's a good practice for helpers.

/cc @Toflar

Reply:
Yeah, lazy loading stuff is good if it benefits the end result. Otherwise it's just confusing. I lack the detailed understanding of this PR here to give a statement about whether or not lazy makes sense here. It certainly doesn't for parameters like `valid_extensions` etc. but it may for others.

Reply:
Then let's inject the parameters directly, so we do not have to use the locator for `parameter_bag`. Then we do not generate PhpStan errors that have to be ignored, either.

Reply:
Like discussed I injected the parameters now but left everything else like it is. 

See e58d09bd802229341f52a302fa4c7946c8e65468